vm/array_test.go

```diff
@@ -140,16 +140,214 @@ func TestArrayIndex(t *testing.T) {
+			a[3, 1]
+		`, []interface{}{4}},
+		{`
+			a = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
+			a[3, 3]
+		`, []interface{}{4, 5, 6}},
+		{`
+			a = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
+			a[3, 5]
+		`, []interface{}{4, 5, 6, 7, 8}},
+		{`
+			a = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
+			a[3, 7]
+		`, []interface{}{4, 5, 6, 7, 8, 9, 10}},
+		{`
+			a = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
+			a[3, 10]
+		`, []interface{}{4, 5, 6, 7, 8, 9, 10}},
+		{`
+			a = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
+			a[-7, 1]
+		`, []interface{}{4}},
+		{`
+			a = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
+			a[-7, 3]
+		`, []interface{}{4, 5, 6}},
+		{`
+			a = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
+			a[-7, 5]
+		`, []interface{}{4, 5, 6, 7, 8}},
+		{`
+			a = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
+			a[-7, 7]
+		`, []interface{}{4, 5, 6, 7, 8, 9, 10}},
+		{`
+			a = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
+			a[-7, 10]
+		`, []interface{}{4, 5, 6, 7, 8, 9, 10}},
+		{`
+			a = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
+			a[2, 3] = [1, 2, 3, 4, 5]
+		`, []interface{}{1, 2, 3, 4, 5}},
+		{`
+			a = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
+			a[2, 3] = [1, 2, 3, 4, 5]
+			a
+		`, []interface{}{1, 2, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10}},
+		{`
+			a = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
+			a[1, 7] = [1, 2, 3]
+			a
+		`, []interface{}{1, 1, 2, 3, 9, 10}},
+		{`
+			a = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
+			a[-5, 3] = [2, 4, 6, 8, 10]
+			a
+		`, []interface{}{1, 2, 3, 4, 5, 2, 4, 6, 8, 10, 9, 10}},
+		{`
+			a = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
+			a[5, 6] = 123
+			a
+		`, []interface{}{1, 2, 3, 4, 5, 123}},
+		{`
+			a = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
+			a[5, 0] = 123
+			a
+		`, []interface{}{1, 2, 3, 4, 5, 123, 6, 7, 8, 9, 10}},
+		{`
+			a = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
+			a[-7, 2] = "@Maxwell-Alexius is solving issue #403"
+			a
+		`, []interface{}{1, 2, 3, "@Maxwell-Alexius is solving issue #403", 6, 7, 8, 9, 10}},
+		{`
+			a = [1, 2, 3, 4, 5]
+			a[5, 123] = 123
+			a
+		`, []interface{}{1, 2, 3, 4, 5, 123}},
+		{`
+			a = [1, 2, 3, 4, 5]
+			a[5, 123] = [1, 2, 3]
+			a
+		`, []interface{}{1, 2, 3, 4, 5, 1, 2, 3}},
+		{`
+			a = [1, 2, 3, 4, 5]
+			a[10, 123] = 123
+			a
+		`, []interface{}{1, 2, 3, 4, 5, nil, nil, nil, nil, nil, 123}},
+		{`
+			a = [1, 2, 3, 4, 5]
+			a[10, 123] = [1, 2, 3]
+			a
+		`, []interface{}{1, 2, 3, 4, 5, nil, nil, nil, nil, nil, 1, 2, 3}},
+		{`
+			a = [1, 2, 3, 4, 5]
+			a[10, 123] = "@Maxwell-Alexius is solving issue #403 which have tons of feature"
+			a
+		`, []interface{}{1, 2, 3, 4, 5, nil, nil, nil, nil, nil, "@Maxwell-Alexius is solving issue #403 which have tons of feature"}},
+	}
+
+	for i, tt := range tests {
+		vm := initTestVM()
+		evaluated := vm.testEval(t, tt.input, getFilename())
+		verifyArrayObject(t, i, evaluated, tt.expected)
+		vm.checkCFP(t, i, 0)
+		vm.checkSP(t, i, 1)
+	}
+}
+
+func TestArrayIndexWithSuccessiveValuesNullCases(t *testing.T) {
+	tests := []struct {
+		input    string
+		expected interface{}
+	}{
+		{`
+			a = [1, 2, 3, 4, 5]
+			a[5, 5] # Range exceeded
```

Comment:
I check the ruby/spec to confirm the ruby's behavior, in this case, ruby returns `[]`.
https://github.com/ruby/spec/blob/875da49e051c030db05cc1a6bcd4a440e9266389/core/array/shared/slice.rb#L52-L53

```
% ruby -ve 'a = [1, 2, 3, 4, 5]; p a[5, 5]'
ruby 2.5.0p0 (2017-12-25 revision 61468) [x86_64-linux]
[]
```

Reply:
Thanks a lot for reminding this special case! <3  i'll change it later on!

Reply:
Double negative can be checked: 

```ruby
a = [1,2,3,4,5]
» a[-1,-1]
#» nil
» a[-1,-1] = 555
#» InternalError: Expect second argument greater than or equal 0. got: -1
```

Zero case:

```ruby
a = [1,2,3,4,5]
#» [1, 2, 3, 4, 5]
» a[1,0]
#» []
» a[1,0] = 555
#» 555
» a
#» [1, 555, 2, 3, 4, 5]